### DISCUSSION TITLE
Why does this not use blender's GLTF exporter

### DISCUSSION THREAD

**Discussion by sdfgeoff** [katharostech/arsenal#116] [Questions] [answered]
Blender has a built in GLTF exporter with support for a huge amount of things (animations, shading, lights etc.), so I'm just wondering why this project wrote it's own GLTF exporter?

**Comment by zicklag** [comment, marked as answer]
Good question! Blender does have a fully featured glTF exporter, but because it's implemented in Python it has a performance disadvantage compared to Arsenal's which is written in Rust.

In a scene with 376,000 vertices made up of 47 subdivided monkey heads, our glTF exporter finishes exporting the entire scene in 12 seconds. In contrast, the official glTF Blender plugin takes 264 seconds to export the same scene on the same computer. That makes our exporter 22 times faster! And that's without even integrating a build cache so that only the models that you change are re-exported.

Other than speed, in the future Arsenal will be probably exporting meshes as GlTF, but will be using a custom scene format to put all the meshes/lights/etc. into the scene for loading by the game engine core. Because of this, we probably won't use the GlTF format for the _entire_ scene and things like lights and other non-mesh objects will be put into the custom scene format that can include all the extra data we need such as scripting information, custom components, etc.

It will take a bit more work on our part to make the exporter, but we really wanted Arsenal to be able to export even large scenes very efficiently without haivng to make you wait any longer than absolutely necessary.

**Reply by sdfgeoff** [reply]
Makes sense.

The reason I stumbled on this project is because I'm also working on a blender -> bevy workflow. If you are curious: my current solution uses a custom file format for meshes, but exports the scene information to bevy's textual "scn" format. I found it easier to do this than to try "append" custom components to the GLTF format.

**Reply by zicklag** [reply]
Yeah, your approach sounds reasonable. Now that you bring it up, I'm not sure how much benefits glTF will actually bring us if we don't use it for scenes. :thinking: There might not be much use for glTF at all because, if I remember correctly, it's not like there's much super revolutionary about the glTF mesh format anyway, other than being a good standard. It's pretty much just JSON plus your vertex buffers, so maybe it would be better just to use a custom mesh format for Arsenal.

Being that your authoratative source file is still the Blender file, which can already export to all kinds of stuff, it doesn't really gain you anything from an inter-op perspective to use glTF, because we're only using it at runtime.

So it sounds like you've got a good idea there. :smile:
